File: CSharp/Wordle.Code/Game.cs
namespace Wordle.Code
{
    public class Game
    {
        private IWordSource wordSource;

        public string Word { get; }

        private int turn = 0;

        public Game(IWordSource wordSource)
        {
            this.wordSource = wordSource;
            this.Word = wordSource.GetWord();
        }

        public GuessResult Guess(string guess)
        {
            if (turn == 6)
            {
                throw new InvalidOperationException("Cannot guess: Out of turns");
            }

            guess = guess.ToUpperInvariant();

            if (string.IsNullOrWhiteSpace(guess) || guess.Length != 5 || !guess.All(l => char.IsLetter(l)))
            {
                throw new ArgumentOutOfRangeException(nameof(guess), "Must be a five letter word");
            }

            if (!this.wordSource.IsValidWord(guess))
            {
                throw new ArgumentException("Not a valid word");
            }

            turn++;
            return new GuessResult(this.Word, guess);
        }
    }
}

File: FSharp/FSharpTest/Game.fs
module Game  

    open System
    open System.Linq

    open WordSource
    open Result

    type Game(source: RandomWordSource) =
        let source = source
        let word = source.GetWord()
        let mutable turn: int = 0

        member this.Turn 
            with get() = turn
            and private set(value) =
                turn <- value

        member this.Guess(guess: string) =
            if turn = 6 then failwith "Out of turns"
            let actualGuess = guess.ToUpperInvariant()

            if String.IsNullOrWhiteSpace(guess) || guess.Length <> 5 || not (guess.All(fun x -> Char.IsLetter(x))) then failwith "Invalid word"
            if not (source.IsValid(guess)) then failwith "Not a valid word"

            this.Turn <- turn + 1
            GuessResult(word, guess)